src/main/java/org/opensearch/ad/task/ADTaskManager.java

```diff
@@ -1538,6 +1544,59 @@ public void updateLatestRealtimeTask(
         }, null, false, listener);
     }
 
+    /**
+     * Update latest realtime task's state, init progress, estimated left minutes and error field.
+     * @param detectorId detector id
+     * @param newState new task state which will override state calculated with rcf total updates
+     * @param rcfTotalUpdates total updates of RCF model
+     * @param detectorIntervalInMinutes detector interval in minutes
+     * @param error error message
+     */
+    public void updateLatestRealtimeTask(
+        String detectorId,
+        String newState,
+        Long rcfTotalUpdates,
+        Long detectorIntervalInMinutes,
+        String error
+    ) {
+        Float initProgress = null;
+        String state = null;
+        // calculate init progress and task state with RCF total updates
+        if (detectorIntervalInMinutes != null && rcfTotalUpdates != null) {
+            state = ADTaskState.INIT.name();
+            if (rcfTotalUpdates < NUM_MIN_SAMPLES) {
+                initProgress = (float) rcfTotalUpdates / NUM_MIN_SAMPLES;
+            } else {
+                state = ADTaskState.RUNNING.name();
+                initProgress = 1.0f;
+            }
+        }
+        // Check if new state is not null and override state calculated with rcf total updates
+        if (newState != null) {
+            state = newState;
+        }
+
+        error = Optional.ofNullable(error).orElse("");
+        if (!adTaskCacheManager.realtimeTaskChanged(detectorId, state, initProgress, error)) {
+            return;
+        }
+        if (ADTaskState.STOPPED.name().equals(state)) {
+            adTaskCacheManager.removeRealtimeTaskCache(detectorId);
+        }
+        Map<String, Object> updatedFields = new HashMap<>();
+        if (initProgress != null) {
+            updatedFields.put(INIT_PROGRESS_FIELD, initProgress);
+            updatedFields.put(ESTIMATED_MINUTES_LEFT_FIELD, Math.max(0, NUM_MIN_SAMPLES - rcfTotalUpdates) * detectorIntervalInMinutes);
+        }
+        if (state != null) {
+            updatedFields.put(STATE_FIELD, state);
+        }
+        if (error != null) {
+            updatedFields.put(ERROR_FIELD, error);
+        }
```

Comment:
What about the case from going to a non-null `error` field to a null `error` field (e.g., user updates a buggy detector and restarts)? Wouldn't we always want to clear the persisted `error` field here?

Reply:
Line 1579 will reset error as empty if no error happens in current interval

`error = Optional.ofNullable(error).orElse("");`

Reply:
I see, makes sense.